### PR TITLE
fix(expr): use checked operations for substr

### LINKED ISSUE #7601: bug: substr panics with out of range arguments
### Describe the bug

```
thread '<unnamed>' panicked at 'byte index 2147483646 is out of bounds of `TEvmWVOycR`', src/expr/src/vector_op/substr.rs:41:23
```

### To Reproduce

```
select substr('a', 2147483646, 1);
```

### Expected behavior
PG:
```
select substr('a', 2147483646, 1);
 substr 
--------
 
(1 row)

select substr('a', 2147483646, 1) = '';
 ?column?
----------
 t
(1 row)
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As per title.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

Closes #7601 
Closes https://github.com/risingwavelabs/risingwave/issues/7632

### CODE REVIEW

File: src/expr/src/vector_op/substr.rs [hunk 1]
```diff
@@ -36,8 +40,12 @@ pub fn substr_start_for(s: &str, start: i32, count: i32, writer: &mut dyn Write)
     if count < 0 {
         bail!("length in substr should be non-negative: {}", count);
     }
-    let begin = max(start - 1, 0) as usize;
-    let end = min(max(start - 1 + count, 0) as usize, s.len());
+    let start = start.checked_sub(1).ok_or(ExprError::NumericOutOfRange)?;
+    let begin = min(max(start, 0) as usize, s.len());
```
Comment (kwannoel) [inline src/expr/src/vector_op/substr.rs:44]: Use this instead of `if block` since that is more friendly for branch prediction?
(I'm not very sure)

Comment (Gun9niR) [reply src/expr/src/vector_op/substr.rs:44]: Rust's slicing behavior is kind of weird IMO 🤣 . `&str[s.len()..]` will return an empty slice, while `&s[s.len() + 1..]` will result in panic. So maybe either short circuit or add some comment to explain the correctness when `begin == s.len()`?

File: e2e_test/batch/functions/substr.slt.part [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# issue: https://github.com/risingwavelabs/risingwave/issues/7632
+statement error out of range
+select substr('W7Jc3Vyufj', (INT '-2147483648'));
```
Comment (xiangjinwu) [inline e2e_test/batch/functions/substr.slt.part:3]: Actually PostgreSQL returns the whole string for this. It does not report NumericOutOfRange when start+count overflows as well.

https://github.com/postgres/postgres/blob/REL_15_1/src/backend/utils/adt/varlena.c#L813

**Review by xiangjinwu — APPROVED** [review]
LGTM, to the extent of changing from panic to ExprError. But the PostgreSQL (and likely SQL standard) behavior is to return empty or whole string in these edge cases.

**Review by Gun9niR — APPROVED** [review]

**Pull request merged**